### PR TITLE
KAFKA-17050: Revert group.version from 3.8

### PR DESCRIPTION
**Pull request by jolshan**
Reverting due to complications when trying to fix KAFKA-17011 in 3.8. Now there will be no production features, so we won't send any over the wire in ApiVersions or BrokerRegistration and cause issues when the receiver is on an old version.

I reverted the typo PR to make the reverts cleaner and minimize chances for errors. The only conflicts were due to imports and a modified test testConsumerGroupDescribe. The fix was to keep the modified parameters but remove the metadataCache code.

### CODE REVIEW

**Comment by jolshan** [comment]
I will run the system tests with this solution as well. 👍

**Comment by cmccabe** [comment]
I looked at `testConsumerGroupDescribe ` and the modified version looks fine to me. If the rest was a clean revert, I can give this a +1.

(I trust you on the import statement conflicts) :)

One question, though, we are not setting `raftSupport` in a bunch of calls to `createKafkaApis` any more. I assume that's not a problem, but it doesn't lok quite correct either (since we're then getting the ZK version of everything.) I suppose if the test doesn't hit any problems we don't need to change it in 3.8 (and it will be fixed in 3.9 and beyond)

**Comment by jolshan** [comment]
@cmccabe Yeah, David and I had a discussion about the raftSupport. Until his change, we were always using the zk metadata cache. However, we are removing the metadata cache from those tests so I don't think it is a big deal. Or I guess to be clear -- we will want to eventually convert to all raft support, but reverting this now leaves us no worse off than before. 

some info: https://github.com/apache/kafka/pull/16120#discussion_r1619225494

**Review by cmccabe — APPROVED** [review]
LGTM when tests pass

File: server-common/src/test/java/org/apache/kafka/server/common/MetadataVersionTest.java [hunk 1]
```diff
@@ -246,8 +243,6 @@ public void testShortVersion() {
         assertEquals("3.7", IBP_3_7_IV2.shortVersion());
         assertEquals("3.7", IBP_3_7_IV3.shortVersion());
         assertEquals("3.7", IBP_3_7_IV4.shortVersion());
-        assertEquals("3.8", IBP_3_8_IV0.shortVersion());
```
Comment (chia7712) [inline server-common/src/test/java/org/apache/kafka/server/common/MetadataVersionTest.java:249]: I guess that we can keep this `assertEquals("3.8", IBP_3_8_IV0.shortVersion());`?

**Review by chia7712 — COMMENTED** [review]
@jolshan thanks for this fix. only one small comment is left. Also, I'm so sorry that the optimized imports cause a trouble on this PR :cry:

**Review by dajac — APPROVED** [review]
LGTM

**Comment by dajac** [comment]
@jolshan Are we going to hit the same issue with `test.feature.version`?

**Comment by jolshan** [comment]
@dajac It is not a production feature so it is not supported by the brokers/controllers.

**Comment by jolshan** [comment]
Kraft upgrade system tests passed. Jenkins failures are unrelated. I will merge.

**Comment by jolshan** [comment]
Just kidding -- I missed @chia7712's comment.

File: server-common/src/test/java/org/apache/kafka/server/common/MetadataVersionTest.java (hunk 1 above)
Comment (jolshan) [reply server-common/src/test/java/org/apache/kafka/server/common/MetadataVersionTest.java:249]: Hmm. It's strange that this was missing before. I can add it back.

**Review by chia7712 — APPROVED** [review]
LGTM

File: core/src/test/scala/integration/kafka/api/IntegrationTestHarness.scala [hunk 2]
```diff
@@ -62,6 +62,7 @@ abstract class IntegrationTestHarness extends KafkaServerTestHarness {
   }
 
   override def generateConfigs: Seq[KafkaConfig] = {
+
```
Comment (junrao) [inline core/src/test/scala/integration/kafka/api/IntegrationTestHarness.scala]: extra newline

File: metadata/src/test/java/org/apache/kafka/controller/QuorumControllerTest.java [hunk 3]
```diff
@@ -298,7 +298,7 @@ public void testFenceMultipleBrokers() throws Throwable {
                     new BrokerRegistrationRequestData().
                         setBrokerId(brokerId).
                         setClusterId(active.clusterId()).
-                        setFeatures(brokerFeatures(MetadataVersion.IBP_3_0_IV1, MetadataVersion.IBP_4_0_IV0)).
+                        setFeatures(brokerFeatures(MetadataVersion.IBP_3_0_IV1, MetadataVersion.IBP_3_7_IV0)).
```
Comment (junrao) [inline metadata/src/test/java/org/apache/kafka/controller/QuorumControllerTest.java]: Hmm, should this be IBP_3_8_IV0?

File: metadata/src/test/java/org/apache/kafka/controller/QuorumFeaturesTest.java [hunk 4]
```diff
@@ -74,7 +74,7 @@ public void testDefaultFeatureMapWithUnstable() {
         for (Features feature : Features.PRODUCTION_FEATURES) {
             expectedFeatures.put(feature.featureName(), VersionRange.of(
                 0,
-                feature.defaultValue(MetadataVersion.latestTesting())
+                feature.defaultValue(MetadataVersion.LATEST_PRODUCTION)
```
Comment (junrao) [inline metadata/src/test/java/org/apache/kafka/controller/QuorumFeaturesTest.java]: Hmm, why is this changed to LATEST_PRODUCTION? The test is about unstable features.

**Review by junrao — COMMENTED** [review]
@jolshan : Thanks for the PR. Left a few comments.

File: core/src/test/scala/integration/kafka/api/IntegrationTestHarness.scala (hunk 2 above)
Comment (jolshan) [reply core/src/test/scala/integration/kafka/api/IntegrationTestHarness.scala]: This wasn't done by me, it was something that David removed in his PR and I reverted it. But sure we can remove it.

File: metadata/src/test/java/org/apache/kafka/controller/QuorumControllerTest.java (hunk 3 above)
Comment (jolshan) [reply metadata/src/test/java/org/apache/kafka/controller/QuorumControllerTest.java]: sure

**Comment by jolshan** [comment]
My goal was to unblock the 3.8 release quickly so I didn't review every line of code I reverted. I can fix these things though.

**Review by junrao — APPROVED** [review]
@jolshan : Thanks for the updated PR. LGTM assuming the tests pass.

**Comment by jolshan** [comment]
Nice -- green build for java 8! And the 3 failures for java 21 were unrelated. Merging now. Thanks all.
